### ISSUE TITLE
M-2100 automation interface protocol

### ISSUE THREAD

**Issue by sfahadshahzad** [jaskie/PlayoutAutomation#12]
Hi,

I am using evertz master switcher which can be controlled using M-2100 protocol. please try to add function to control device using the same protocol.

Regards,

Fahad

**Comment by jaskie** [comment]
@sfahadshahzad: What specific functionality do you expect?
Regards,
Jurek

**Comment by sfahadshahzad** [comment]
i thought if MCR (Master Control Room) switcher is controllable from playlist application so we can use lots of  functions like switching on different source, we can also use event based logo's which are already available in switcher.

Down stream Keyer feature also available if required can be controlled.

Regards,

Fahad

**Comment by jaskie** [comment]
So we have:
- Input switching.
There is a plan to support source selection for Nevion routers, so all concept and UI work will be then done.
- Logo switching
Logo and parental signs control is available by IGpi interface, so it's relatively easy to support other protocols as well (and it would already work with any device having GPI interface using [Advantech GPI card](http://www.advantech.com/products/1-2mlkb0/pcie-1730/mod_6d7a4820-cf93-4f6b-bc3a-09d4e1ca62bc)).
- DSK: currently I do not see any specific usage.

Moreover, supporting a MCR switcher is not my aim, because the system can replace it thoroughly, with keying, channel branding and live event broadcasting. Even special effects, such a two-window view can be achieved using built-in script functionality. Only first thing (input switching) is in plans, but rather for video routers (mentioned Nevion), as input pre-selection functionality. We also do not have a spare switcher to test such a integration.

@sfahadshahzad: you can order such integration to an developer or do this yourself, I will help on this system's side and willingly add the code to main branch.
Regards,
Jurek

**Comment by theusmpa** [comment]
Hello @jaskie, 

I am also very interested that the application controls an MCR table, it would be something like: we would have the live events entering the MCR and the servers would also enter the MCR, thus configuring the media events to control an entry and the other live events (as it is now), basically we would need to configure the media events to assume an entry in the MCR and switch as soon as the play is done, the same for when a Live event is displayed (it would have to be immediate and not 2 seconds before as it is currently done).

The PGM signal would be the output of the MCR, so if we have a crash in the playout server, we would have the option of putting LIVE on the air.

In my case, 80% of the programming is satellite (live), so the server would be to run the commercial load, in the end it would have a much more robust system and without the need for the PGM signal to pass through a server.

Our current automation, based on ADC (provided by former Harris), works that way.

It could even be a BMD (as we already have the code written to control a BMD, it would only be a matter of adapting and adding this type of configuration to the application)

Do you believe it is possible to add this feature?

Thank you.

**Comment by jaskie** [comment]
@theusmpa, I just have read your post third time, and still didn't grasp what is going on. Would you like to clarify, maybe write in other words, or add a diagram? My English skills a somewhat limited...

**Comment by m1tk4** [comment]
Let me try to translate from "broadcast":

1. I think what @theusmpa wants is that during the live segments the playout automation would change the routing on the switcher so it *bypasses* the playout system completely. Effectively, external router control events at the beginning of live segments rather than controlling internal video pipeline.

2. BMD is likely Blackmagic Design, something like their ATEM switcher.

**Comment by theusmpa** [comment]
Hello friends!

I apologize for not making me understand, but I will try to explain using two images.

In this first image I am illustrating the topology, which, in my opinion, would be the current solution, that is, I control a switcher that enters casparCG to finally deliver the final signal (I call it PGM).
So I have my rundown with the media and the Live events that will select the inputs of this switcher two seconds so that it will be displayed on my PGM signal at the appropriate time.

![img1](https://user-images.githubusercontent.com/6371447/85262619-685e2e80-b444-11ea-9a43-70ef4980f6d7.png)


My need is illustrated in the image below.
Basically I need automation to control the switcher, so the server will serve as an input source.

![img2](https://user-images.githubusercontent.com/6371447/85262644-72802d00-b444-11ea-9641-ea5680cad761.png)


My suggestion would be: set up so that media events have the ability to control a switcher source, just like live events do, and this switching happens at the exact moment it is on the air, and not two seconds before being displayed as happens in the moment.

Would this be possible to implement in the near future?

Thank you.

**Comment by jaskie** [comment]
Ok, thanks for clarification. It seems to be reasonable, IMHO it should be added a "delay" configurable property for switch time (+- 2 seconds from event start time). It not be exact "zero", because the output from CasparCG is delayed by at least 4 frames, and you'll need to compensate it.
Last thing I do not understand is why apply routing to other than live events? Shouldn't rather be Live on/ Live off properties? Or even a global setting for all movie items? It also need ability to be disabled, when you want to start a rundown during live for testing the rundown.

**Comment by theusmpa** [comment]
In the case of the media it would be a global configuration, for live events it would be a unique configuration for each event, I will try to exemplify: imagine a live event called SP, another live event called BH and another live event called REDE, each one of them would receive a value unique referring to the wired input on the switcher, and the media would also receive a value (the same value for all media, since the source that generates this content and the casparCG server) referring to its respective entry on the switcher. That way it would allow me to assemble according to the second image above.

It would be a very robust solution, since the main signal will not be passing through the server all the time, only when the media, which in my case are local commercials, were being displayed.

In this topology, the cut must be immediate, both the Live event and the media, because whoever is actually making the cut and the switcher, we would not have any delay problems or anything like that, and the main output of the signal would be one of the outputs of the switcher and no longer the server output.

It is this concept that we currently use with proprietary, expensive and limited solutions from third parties that I intend to replace with your application, but I need to worry about robustness, because 80% of my programming is via live event (in my case a satellite we call REDE) .

Using this concept, I agree with you that there should be a button, or something like that, on the main screen that would make it possible to activate and deactivate this control of the switcher, in order not to happen undue activations when performing a test or even when a backup instance is being used, avoiding the switcher to receive double cut command (but as it will happen at the same time for the same input, I believe that there will be no problems, we need to test it).

Does that make sense to you now?

I may also be mistaken and have misunderstood the current working concept of your solution.

What do you think?

Thank you

Matheus
